### PR TITLE
[Refactor][meta-format-change] Change image header to json format

### PR DESCRIPTION
**Pull request by gengjun-git**
## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #

## Problem Summary(Required):

1. Change the image header to json format.
2. Put an integer (starrocksMetaVersion) on the begin of image, so that we can distinguish the json-format and old format.
3. The current version format is [-1, metaVersion, starrocksMetaVersion], but the metaVersion is deprecated, so we change the format to [starrocksMetaVersion], and we can use the first int to distinguish the two formats.
4. Remove all deprecated metaVersion related code.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java [hunk 1]
```diff
@@ -43,9 +43,9 @@ public class FeConstants {
     // bloom filter false positive probability
     public static final double DEFAULT_BLOOM_FILTER_FPP = 0.05;
     // general model
-    // Current metadata version. Use this version to write journals and image
-    // for community meta version
-    public static final int META_VERSION = FeMetaVersion.VERSION_CURRENT;
+    @Deprecated
+    // for rollback compatible
+    public static final int META_VERSION = 92;
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java:48]: can't we just remove this member?

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 2]
```diff
@@ -1501,68 +1497,71 @@ private void processMvRelatedMeta() {
...
-            }
-            MetaContext.get().setMetaVersion(metaVersion);
+            // when flag is positive, this is new version format
+            starrocksMetaVersion = flag;
         }
 
-        long replayedJournalId = dis.readLong();
-        newChecksum ^= replayedJournalId;
+        if (starrocksMetaVersion > FeConstants.STARROCKS_META_VERSION) {
+            LOG.error("invalid meta data version found, cat not bigger than FeConstants.starrocks_meta_version."
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1517]: typo, and the logging message is also not proper, if `starrocksMetaVersion > FeConstants.STARROCKS_META_VERSION`, it means that  we use an old version FE to load newer version image, what we should do is to use right(newer) version FE to restart, not just modify the code.

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 3]
```diff
@@ -1501,68 +1497,71 @@ private void processMvRelatedMeta() {
...
+
+    public long loadHeaderV2(DataInputStream dis, long checksum) throws IOException {
+        ImageHeader header = GsonUtils.GSON.fromJson(Text.readString(dis), ImageHeader.class);
+
+        checksum ^= header.getBatchEndId();
+        idGenerator.setId(header.getBatchEndId());
+
+        isDefaultClusterCreated = header.isDefaultClusterCreated();
+
+        LOG.info("finished replay header from image");
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1557]: finished to

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 4]
```diff
@@ -1501,68 +1497,71 @@ private void processMvRelatedMeta() {
...
+
+        isDefaultClusterCreated = dis.readBoolean();
 
         LOG.info("finished replay header from image");
-        return newChecksum;
+        return checksum;
+    }
+
+    public long loadHeaderV2(DataInputStream dis, long checksum) throws IOException {
+        ImageHeader header = GsonUtils.GSON.fromJson(Text.readString(dis), ImageHeader.class);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1550]: we don't need `replayedJournalId` here any more?

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java (hunk 2 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1517]: does this `if` condition here means that we cannot rollback to older version, like from 3.1 to 3.0?

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 5]
```diff
@@ -1790,6 +1791,20 @@ public long saveHeader(DataOutputStream dos, long replayedJournalId, long checks
         return checksum;
     }
 
+    // TODO [meta-format-change]
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1794]: meaning this pr need to go to 3.0 first?

File: fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java (hunk 1 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java:48]: not, we should write this version in 3.0 for compatible.

**Review by HangyuanLiu — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java (hunk 4 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1550]: yes, it's useless now.

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java (hunk 2 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java]: changed

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java (hunk 5 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1794]: methods marked with [meta-format-change] will be used in 3.1.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 195 / 447 (43.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/EsTable.java|0|38|00.00%|[351, 352, 372, 373, 374, 375, 376, 378, 379, 380, 381, 382, 383, 384, 385, 387, 388, 389, 393, 394, 395, 397, 399, 401, 402, 403, 406, 407, 409, 411, 412, 414, 417, 418, 419, 420, 421, 423]|
|:large_blue_circle:|com/starrocks/backup/Repository.java|0|1|00.00%|[380]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadManager.java|0|1|00.00%|[685]|
|:large_blue_circle:|com/starrocks/catalog/HiveTable.java|0|1|00.00%|[375]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadManager.java|0|1|00.00%|[737]|
|:large_blue_circle:|com/starrocks/persist/DatabaseInfo.java|0|4|00.00%|[107, 109, 111, 112]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|0|1|00.00%|[874]|
|:large_blue_circle:|com/starrocks/load/BrokerFileGroup.java|0|4|00.00%|[495, 496, 499, 500]|
|:large_blue_circle:|com/starrocks/analysis/LabelName.java|0|1|00.00%|[119]|
|:large_blue_circle:|com/starrocks/load/DeleteInfo.java|0|9|00.00%|[177, 178, 180, 181, 182, 183, 186, 188, 189]|
|:large_blue_circle:|com/starrocks/transaction/GlobalTransactionMgr.java|0|5|00.00%|[760, 761, 762, 763, 764]|
|:large_blue_circle:|com/starrocks/load/DppConfig.java|0|3|00.00%|[404, 414, 425]|
|:large_blue_circle:|com/starrocks/backup/BackupHandler.java|0|1|00.00%|[668]|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|0|22|00.00%|[523, 524, 525, 526, 527, 530, 531, 532, 533, 536, 1143, 1144, 1145, 1146, 1147, 1148, 1149, 1150, 1151, 1152, 1154, 1156]|
|:large_blue_circle:|com/starrocks/persist/PartitionPersistInfo.java|0|2|00.00%|[136, 137]|
|:large_blue_circle:|com/starrocks/catalog/ColocateTableIndex.java|0|25|00.00%|[681, 682, 683, 684, 685, 686, 687, 688, 689, 691, 692, 694, 695, 696, 697, 698, 699, 700, 701, 703, 705, 708, 709, 710, 731]|
|:large_blue_circle:|com/starrocks/common/util/SmallFileMgr.java|0|1|00.00%|[545]|
|:large_blue_circle:|com/starrocks/persist/TablePropertyInfo.java|0|2|00.00%|[98, 99]|
|:large_blue_circle:|com/starrocks/transaction/PartitionCommitInfo.java|0|2|00.00%|[109, 110]|
|:large_blue_circle:|com/starrocks/persist/ColocatePersistInfo.java|0|1|00.00%|[118]|
|:large_blue_circle:|com/starrocks/persist/ModifyPartitionInfo.java|0|1|00.00%|[155]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|4|00.00%|[498, 499, 500, 501]|
|:large_blue_circle:|com/starrocks/load/Load.java|0|9|00.00%|[1081, 1082, 1083, 1086, 1087, 1088, 1091, 1092, 1093]|
|:large_blue_circle:|com/starrocks/system/Frontend.java|0|1|00.00%|[198]|
|:large_blue_circle:|com/starrocks/persist/TruncateTableInfo.java|0|2|00.00%|[88, 89]|
|:large_blue_circle:|com/starrocks/qe/VariableMgr.java|0|3|00.00%|[337, 338, 581]|
|:large_blue_circle:|com/starrocks/analysis/TableRef.java|0|1|00.00%|[419]|
|:large_blue_circle:|com/starrocks/load/loadv2/BulkLoadJob.java|0|10|00.00%|[339, 344, 345, 346, 347, 348, 350, 351, 353, 354]|
|:large_blue_circle:|com/starrocks/catalog/DiskInfo.java|0|2|00.00%|[199, 200]|
|:large_blue_circle:|com/starrocks/plugin/PluginMgr.java|0|1|00.00%|[348]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|0|9|00.00%|[1213, 1214, 1215, 1216, 1217, 1218, 1221, 1288, 1289]|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|0|1|00.00%|[826]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|1|6|16.67%|[1178, 1179, 1181, 1185, 1186]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|20|20.00%|[3878, 3879, 3880, 3881, 3882, 3884, 3885, 3890, 3892, 3896, 3897, 3899, 3901, 3906, 3907, 3908]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|2|6|33.33%|[587, 588, 589, 590]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|10|40.00%|[749, 750, 751, 752, 753, 756]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|1|2|50.00%|[1971]|
|:large_blue_circle:|com/starrocks/mysql/privilege/UserProperty.java|5|9|55.56%|[258, 264, 265, 266]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|34|61|55.74%|[1371, 1513, 1517, 1518, 1519, 1564, 1588, 1589, 1591, 1597, 1598, 1599, 1600, 1601, 1602, 1603, 1605, 1610, 1611, 1612, 1615, 1623, 1643, 1702, 1774, 1775, 1776]|
|:large_blue_circle:|com/starrocks/load/ExportJob.java|4|7|57.14%|[925, 926, 927]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|8|13|61.54%|[1535, 1536, 1563, 1564, 1567]|
|:large_blue_circle:|com/starrocks/persist/PrivInfo.java|2|3|66.67%|[188]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|8|11|72.73%|[1616, 1617, 1618]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|6|8|75.00%|[430, 431]|
|:large_blue_circle:|com/starrocks/system/Backend.java|11|14|78.57%|[350, 351, 352]|
|:large_blue_circle:|com/starrocks/catalog/MetaVersion.java|6|7|85.71%|[64]|
|:large_blue_circle:|com/starrocks/persist/ReplicaPersistInfo.java|7|8|87.50%|[353]|
|:large_blue_circle:|com/starrocks/transaction/TransactionState.java|8|9|88.89%|[813]|
|:large_blue_circle:|com/starrocks/catalog/Table.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/ExportMgr.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/ImageHeader.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/DropInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/loadv2/LoadJob.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/LocalTablet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/PrivEntry.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FsBroker.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TempPartitions.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/MysqlTable.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/DropPartitionInfo.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/DropDbInfo.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/GlobalPrivEntry.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/CreateTableInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/WhiteList.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/MetaObject.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Column.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Replica.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Role.java|5|5|100.00%|[]|

**Review by nshangyiming — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**